CODING_GUIDELINES.md

```diff
@@ -151,6 +151,55 @@ prior to attempting a search.
 Note: We are in the process of replacing methods that don't adhere to the above principles (for example, return a
 structure instead of a pointer, return a nil error on not-found, etc).
 
+## Implementing functions to support different API versions
+
+Functions dealing with different versions should use a matrix structure to identify which calls to run according to the 
+highest API version supported by vCD. An example can be found in adminvdc.go.
+
+```
+type vdcProducer struct {
```

Comment:
In the light of possible design patterns, I think we are misnaming this. Maybe I'm missing something here and if that's the case please share _a link_ to the design pattern that matches current implementation, but my reasoning is:

a.) Yes, there is a **producer**-consumer pattern, but it is used for **multi-threading** use cases where producer is generating data and consumer - consuming it. E.g.:
https://tpaschalis.github.io/golang-producer-consumer/

b.) Instead, what we are doing here is returning specific function implementations depending on the vCD version, which in my view much better maps to the **factory** pattern. E.g. (though not specifically our use case):
https://www.sohamkamani.com/blog/golang/2018-06-20-golang-factory-patterns/
https://travix.io/factory-method-pattern-in-go-2e09b233453e

Hence, what about naming all these a `factory` as opposed to `producer`? Or any better pattern name that we are actually using here?

CC: @dataclouder @Didainius

Reply:
If we are to be picky with names I would say this is a lot simpler than some factory because it is tied to `vdc` handling only. This is just struct with function definitions for specific version and the `var vdcProducerByVcdVersion = map[string]vdcProducer` is just a map holding it.


I could go with something like 
`type vdcVersionCrud struct`
`type vdcVersionedFunctions struct`
`type vdcFunctionsByVersion struct`

and name of the map like
`var vdcVersionMapCrud = map[string]vdcVersionedCrud`

Reply:
> because it is tied to vdc handling only

Well, this scheme will incrementally grow into all other files as well wherever we'll need to add a feature available in the newer versions of vCD. So, let's pick the best matching name for this producer/factory/map :)

To add to your suggestions:
`vdcCrudByVcdVersion`
But I don't really like it, because it is likely that such map may contain not only CRUD functions. The fact that we already have a function `validateVdcConfigurationV97` is a good indication of that.

Reply:
I don't agree that `validateVdcConfigurationV97` is good example. Mapping is for main points main functions mapping. Validation it's inside info. I don't expect them to go to mapping as we may create spaghetti in that way very soon.

Reply:
@vbauzysvmware Also one more question about the approach. Right now you're "overloading" only Create and Update functions. How would it look like in code if you'd need to "overload" some Get functions as well? Imagine you need to add multi-version support for `adminorg.go`.

Reply:
It would like as it is now. It's easy to add get, delete or any function. You just enhance interface, add mappings and change main function get/delete to call private functions by version which are defined in already mentioned maps. It's 2 cents todo that.

Reply:
> > because it is tied to vdc handling only
> 
> Well, this scheme will incrementally grow into all other files as well wherever we'll need to add a feature available in the newer versions of vCD. So, let's pick the best matching name for this producer/factory/map :)
> 
> To add to your suggestions:
> `vdcCrudByVcdVersion`
> But I don't really like it, because it is likely that such map may contain not only CRUD functions. The fact that we already have a function `validateVdcConfigurationV97` is a good indication of that.

Also keep in mind, that aim is to have separate maps for different types. So these maps are only for vDC functions. If we have specific functions to need mapping, then we will now how to name it.

Reply:
@vbauzysvmware Will it always be exclusive to CRUD functions or any function may land into this map?

Reply:
For this moment I don't see any other function which needs such thing unless CRUD

Reply:
Then a few suggestions based on all of the above:

a.)
![Screenshot 2020-02-10 at 16 45 05](https://user-images.githubusercontent.com/11330689/74159586-b35d6f00-4c24-11ea-957a-9f14455649ef.png)

b.)
![Screenshot 2020-02-10 at 16 40 42](https://user-images.githubusercontent.com/11330689/74159629-bfe1c780-4c24-11ea-945d-231f99d11b67.png)

Thoughts?

Reply:
a). It gives a hint that this is related to versions.

Reply:
For me `a)` better. P.s. as matter of fact, current `producer` was inclined and agreed with Giuseppe.

Reply:
I believe this last choice, though longer, is clearer, because:
- Just by looking at the var name `vdcVersionedFuncs...` you know that it is about versioning.
- The names of all the four vars are consistent among each other.
- The fact that we have "version" in the names make them handily inline with related var names of `api_vcd_versions.go` (`VersionInfo`, `SupportedVersions`, `apiVersionToVcdVersion`, etc.)
- It doesn't imply the producer-consumer multi-threading pattern.

Reply:
done

Reply:
Thanks, two comments:
1.) Proposal was plural `Funcs` but you made singular `Func` - was this on purpose?
2.) `producer` is still used as var across the files, it should be changed to `vdcFuncs` or something like that.

Reply:
done